File: 283. Move Zeroes/main.cpp
#include <vector>

using std::vector;
using std::swap;

class Solution {
public:
    void moveZeroes(vector<int> &nums) {
        size_t non_zero_index = 0;

        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] != 0) {
                swap(nums[non_zero_index], nums[i]);
                non_zero_index++;
            }
        }
    }
};

class Solution {
public:
    void moveZeroes(vector<int> &nums) {
        size_t zero_pos = 0;

        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] != 0) {
                nums[zero_pos] = nums[i];
                zero_pos++;
            }
        }

        for (int i = zero_pos; i < nums.size(); i++) {
            nums[i] = 0;
        }
    }
};

class Solution {
public:
    void moveZeroes(vector<int> &nums) {
        int index = 0;

        for (int i = 0; i < nums.size(); i++) {
            if (nums[index] == 0) {
                nums.erase(nums.begin() + index);
                nums.push_back(0);
            } else {
                index++;
            }
        }
    }
};

File: 283. Move Zeroes/main.rs
impl Solution {
    pub fn move_zeroes(nums: &mut Vec<i32>) {
        let count = nums.iter().filter(|&&x| x == 0).count();
        nums.retain(|&x| x != 0);
        nums.extend(std::iter::repeat(0).take(count));
    }
}

impl Solution {
    pub fn move_zeroes(nums: &mut Vec<i32>) {
        let mut non_zero_index = 0;

        for i in 0..nums.len() {
            if nums[i] != 0 {
                nums.swap(non_zero_index, i);
                non_zero_index += 1;
            }
        }
    }
}

impl Solution {
    pub fn move_zeroes(nums: &mut Vec<i32>) {
        let mut zero_pos = 0;

        for i in 0..nums.len() {
            if nums[i] != 0 {
                nums[zero_pos] = nums[i];
                zero_pos += 1;
            }
        }

        for i in zero_pos..nums.len() {
            nums[i] = 0;
        }
    }
}
